### PR TITLE
File upload improvements and tests

### LINKED ISSUE #3744: Upload attributes an upload to the system user, it should credit the user who uploads
If you have a simple file that you want to upload the user given credit for any changes is the system user, thus an offline translators won't get the credit for their work.

### PR DESCRIPTION
**Pull request by phlax**
When translation files are uploaded any changes are accredited to the correct
user, rather than the system user.

When user uploads a translation file any changes are recorded with the correct
submission type

Fixes #3744, #3745, #3977

### CODE REVIEW

**Comment by dwaynebailey** [comment]
Looks good, I'd like to test first and get a GTM from others.  If I read correctly this doesn't address the submisson type does it?

**Comment by phlax** [comment]
@dwaynebailey no - that is #3745 i think

**Comment by phlax** [comment]
k - ive added a commit to address the submission type issue too.

Ive included both in one PR as they would otherwise conflict

**Comment by julen** [comment]
Looks good, however I'd expect the tests to be testing for import/export's `import_file()` method, not `store.update()`.

**Comment by phlax** [comment]
there are tests for both

**Comment by julen** [comment]
Can you point me where in this PR?

**Comment by phlax** [comment]
in tests/import_export/import.py

I figured the initial file handling and raising of errors should be tested in tests/import_export/import.py and the tests for the store update with a filestore should be in the store tests

**Comment by julen** [comment]
There's no such file, it must be in your local repo only.

**Comment by phlax** [comment]
there is in that PR 

https://github.com/translate/pootle/pull/3979/files

**Comment by julen** [comment]
Sorry I haven't gone through all my notifications yet and as you might imagine it's unexpected to have them in a separate PR. Also, unless we're talking about a huge invasive feature branch, it'd help to keep PRs focused on a single issue to ease reviews.

**Comment by phlax** [comment]
Mostly I have kept separate bugs/commits to separate PRs

The referenced PR (#3979) does a different thing - it tests the import_files function.

The tests added here test the store.update method with particular types of storefile

**Comment by unho** [comment]
I do agree with @julen that having two separate PRs that propose intertwined tests does make more difficult to review them.

File: pootle/apps/import_export/utils.py [hunk 1]
```diff
@@ -12,10 +12,14 @@
...
 from pootle_store.models import Store
+from pootle_statistics.models import SubmissionTypes
 
 
-def import_file(file):
+def import_file(file, user=None):
     f = getclass(file)(file.read())
+    if not hasattr(f, "parseheader"):
+        raise ValueError(_("Unsupported filetype '%s', only po files " % (file.name)
+                           + "are supported at this time\n"))
```
Comment (unho) [inline pootle/apps/import_export/utils.py]: Please do not do this. Instead use:

```python
        raise ValueError(_("Unsupported filetype '%s', only PO files are "
                           "supported at this time.\n" % file.name))
```

Also I don't think that `file.name` actually provides the filetype.

File: pootle/apps/pootle_store/models.py [hunk 2]
```diff
@@ -1610,16 +1610,26 @@ def remove_modified_units(self, unit_ids):
...
-                     revision=None):
+                     revision=None, submission_type=None):
         """Updates existing units in the store.
 
         :param uids_to_update: UIDs of the units to be updated.
         :param uid_index_map: dictionary of DB ID to index mappings.
         :param user: attribute specific changes to this user.
         :param revision: set updated unit revision to this value.
-        :return: The number of units that were actually updated.
+        :param submission_type: set submission type for update.
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: The line above is added in the wrong commit.

File: pootle/apps/pootle_store/models.py [hunk 3]
```diff
@@ -1720,6 +1745,8 @@ def update(self, overwrite=False, store=None, only_newer=False):
             the current DB store's FS store will be used as a source.
         :param only_newer: Whether to update only the files that changed on
             disk after the last sync.
+        :param user: User to attribute updates to.
+        :param submission_type: Submission type of saved updates.
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: The changes above are in the wrong commit.

File: pootle/apps/import_export/utils.py (hunk 1 above)
Comment (phlax) [inline pootle/apps/import_export/utils.py]: get_class has responsibility for parsing the mimetype - file.name is just used to indicate to the user the name of the file that they have uploaded

File: pootle/apps/pootle_store/models.py (hunk 3 above)
Comment (phlax) [inline pootle/apps/pootle_store/models.py]: yep, my bad - but to go back now and put it into the correct commit is not trivial - apologies

**Comment by phlax** [comment]
@unho i have merged the commits into this PR

File: pootle/apps/import_export/utils.py (hunk 1 above)
Comment (unho) [inline pootle/apps/import_export/utils.py]: Then I suggest you to rephrase the warning message to: Unsupported filetype **for** '%s'

File: pootle/apps/pootle_store/models.py (hunk 3 above)
Comment (unho) [inline pootle/apps/pootle_store/models.py]: It is easy but takes some time. Just do an interactive rebase for the commit where the changes should go, add the changes to the commit and finish interactive rebase. Then force push the PR branch.

Comment (phlax) [inline pootle/apps/pootle_store/models.py]: im sorry @unho im running out of time

File: tests/import_export/import.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+def test_import_success(en_tutorial_po):
+    _import_file(IMPORT_SUCCESS)
+
+
+@pytest.mark.django_db
+def test_import_failure(file_import_failure, en_tutorial_po):
+    filename, response = file_import_failure
+    with pytest.raises(ValueError) as e:
+        _import_file(filename)
+    assert e.value.message == response % filename
```
Comment (unho) [inline tests/import_export/import.py]: Forgive me for my ignorance on how pytest works, but how does this test all the cases?

File: tests/import_export/import.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                                       f.read(),
+                                       content_type))
+
+
+@pytest.mark.django_db
+def test_import_success(en_tutorial_po):
+    _import_file(IMPORT_SUCCESS)
+
+
+@pytest.mark.django_db
```
Comment (unho) [inline tests/import_export/import.py]: Is it necessary to use `@pytest.mark.django_db` if this is not expected to hit the database?

File: tests/import_export/import.py (hunk 4 above)
Comment (unho) [inline tests/import_export/import.py]: I honestly would prefer to explicitly test case by case.

File: pootle/apps/pootle_store/models.py [hunk 6]
```diff
@@ -1610,16 +1610,27 @@ def remove_modified_units(self, unit_ids):
...
+            is the number of suggestions added due to revision conflicts.
         """
         updated = 0
+        suggested = 0
+
+        store_revision = None
+        if hasattr(store, "parseheader"):
+            pootle_revision = store.parseheader().get("X-Pootle-Revision",
+                                                      None)
+            if pootle_revision:
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: Better use:

```python
            if pootle_revision is not None:
```

File: pootle/apps/pootle_store/models.py [hunk 7]
```diff
@@ -1753,6 +1781,13 @@ def update(self, overwrite=False, store=None, only_newer=False):
         self.state = LOCKED
         self.save()
 
+        store_revision = None
+        if hasattr(store, "parseheader"):
+            pootle_revision = store.parseheader().get("X-Pootle-Revision",
+                                                      None)
+            if pootle_revision:
+                store_revision = int(pootle_revision)
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: This is the second appearance of this block of code. Please refactor to a method.

File: pootle/apps/pootle_store/models.py [hunk 8]
```diff
@@ -1794,9 +1830,14 @@ def insert(i1, j1, j2, offset):
...
                     uid = unit.getid()
                     if uid not in old_unitid_set:
-                        self.addunit(unit, new_unit_index, user=system,
-                                     revision=update_revision)
-                        changes['added'] += 1
+                        # Don't add unit if store revision is less than max
+                        # unit revision
+                        add_unit = (store_revision is None
+                                    or max_unit_revision <= store_revision)
+                        if add_unit:
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: The comment already states what the condition does. I think it is ok to directly use the condition instead of creating a temporary variable.

File: tests/import_export/import.py (hunk 4 above)
Comment (unho) [inline tests/import_export/import.py]: I have read about parametrizing a fixture https://pytest.org/latest/fixture.html#parametrizing-a-fixture I am now ok with keeping this code as long as we add some explanatory comment. Tests shouldn't be difficult to understand.

File: pootle/apps/pootle_store/models.py (hunk 3 above)
Comment (phlax) [inline pootle/apps/pootle_store/models.py]: these changes have been put into the correct commits

File: tests/import_export/import.py (hunk 5 above)
Comment (phlax) [inline tests/import_export/import.py]: I think so as you want the tutorial store to be generated

File: pootle/apps/pootle_store/models.py (hunk 6 above)
Comment (phlax) [inline pootle/apps/pootle_store/models.py]: done

File: pootle/apps/pootle_store/models.py (hunk 7 above)
Comment (phlax) [inline pootle/apps/pootle_store/models.py]: I refactored to util function as it would be unclear what it did on the model

**Comment by phlax** [comment]
@unho ive put the changes into the correct commits and made other changes that you have suggested

I also added some custom exceptions for file imports

File: pootle/apps/pootle_store/models.py (hunk 2 above)
Comment (unho) [inline pootle/apps/pootle_store/models.py]: Now fixed.

File: pootle/apps/pootle_store/models.py [hunk 9]
```diff
@@ -658,7 +659,7 @@ def sync(self, unit):
 
     def update(self, unit, user=None):
         """Update in-DB translation from the given :param:`unit`.
-
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: Please keep this blank line.

File: pootle/apps/pootle_store/models.py [hunk 10]
```diff
@@ -658,7 +659,7 @@ def sync(self, unit):
...
+        :param user: User to attribute updates to.
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: This is an unrelated cleanup, so please create a separate commit for it and directly merge it to master.

File: pootle/apps/pootle_store/models.py [hunk 11]
```diff
@@ -1616,6 +1622,17 @@ def update_units(self, store, uids_to_update, uid_index_map, user,
             uid = unit.getid()
             newunit = store.findid(uid)
 
+            # If the unit's revision is greater than the store's add a
+            # suggestion instead
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: Please add a period at the end.

File: pootle/apps/pootle_store/models.py [hunk 12]
```diff
@@ -1780,9 +1806,14 @@ def insert(i1, j1, j2, offset):
                     new_unit_index = i1 + index + 1 + offset
                     uid = unit.getid()
                     if uid not in old_unitid_set:
-                        self.addunit(unit, new_unit_index, user=system,
-                                     revision=update_revision)
-                        changes['added'] += 1
+                        # Don't add unit if store revision is set and is less
+                        # than max unit revision
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: Final period please.

File: pootle/apps/pootle_store/util.py [hunk 13]
```diff
@@ -118,3 +118,11 @@ def get_change_str(changes):
...
 
     return "no changed"
+
+
+def parse_pootle_revision(store):
+    if hasattr(store, "parseheader"):
+        pootle_revision = store.parseheader().get("X-Pootle-Revision",
+                                                  None)
+        if pootle_revision is not None:
+            return int(pootle_revision)
```
Comment (unho) [inline pootle/apps/pootle_store/util.py]: This doesn't return a default value.

Comment (phlax) [inline pootle/apps/pootle_store/util.py]: No its not supposed to

File: pootle/apps/import_export/utils.py [hunk 14]
```diff
@@ -12,35 +12,32 @@
 from django.utils.translation import ugettext as _
 
 from pootle_store.models import Store
+from pootle_statistics.models import SubmissionTypes
 
+from .exceptions import (UnsupportedFiletypeError, MissingPootlePathError,
+                         MissingPootleRevError, FileImportError,
+                         ERR_UNSUPPORTED_FILETYPE, ERR_MISSING_POOTLE_PATH,
+                         ERR_MISSING_POOTLE_REV, ERR_FILE_IMPORT)
```
Comment (unho) [inline pootle/apps/import_export/utils.py]: According to our styleguide http://docs.translatehouse.org/projects/translate-toolkit/en/latest/developers/styleguide.html#imports the constants go first.

File: pootle/apps/pootle_store/util.py (hunk 13 above)
Comment (phlax) [inline pootle/apps/pootle_store/util.py]: however i dont think we want to replace the not isinstance(store_revision, int) as the rest of the condition relies on the first value being an int

**Comment by phlax** [comment]
@unho i think this is good to go

File: pootle/apps/import_export/utils.py [hunk 15]
```diff
@@ -12,35 +12,32 @@
...
+
+def import_file(file, user=None):
     f = getclass(file)(file.read())
+    if not hasattr(f, "parseheader"):
+        raise UnsupportedFiletypeError(_(ERR_UNSUPPORTED_FILETYPE) % file.name)
     header = f.parseheader()
     pootle_path = header.get("X-Pootle-Path")
     if not pootle_path:
-        raise ValueError(_("File '%s' missing X-Pootle-Path header\n") % (file.name))
+        raise MissingPootlePathError(_(ERR_MISSING_POOTLE_PATH) % file.name)
```
Comment (julen) [inline pootle/apps/import_export/utils.py]: This doesn't work as already outlined in another PR. Note the actual error messages belong here, i.e. you want to write the actual message wherever you raise the exceptions.

I see you have added these to check for exception messages in tests. However, since you already added specialized exceptions, I believe we are already covered.

Comment (phlax) [inline pootle/apps/import_export/utils.py]: partly, but also because i usually try to keep strings in constants so that they are easier to change/reuse.
